### DISCUSSION TITLE
Equidistant ticks on axis

### DISCUSSION THREAD

**Discussion by daniel-eh** [patrykandpatrick/vico#970] [Questions]
### Question

Hello.

I'm looking to place ticks equidistant-ish from each other on the axis. So for a simple example, if I have 100 x values, I want to plot them all but I only want a tick on the x-axis at every 20 (20, 40, 60, 80). I don't necessarily know how many data points I'll get, so I'd like to be able to essentially just pass an argument for how many ticks go in the axis, evenly spread. There's a picture attached to sort of demonstrate this. 

I suspect I'm supposed to achieve this via `xStep` but I've been struggling to get it to work/look the way I need it to.

Thanks for any help!

<img width="336" alt="Screenshot 2024-12-10 at 1 36 09 PM" src="https://github.com/user-attachments/assets/d36df50c-4c28-4cff-8dfb-fd83f693b88b">

### Vico version(s)

2.0.0-beta.3

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Hello, @daniel-eh. If the tick logic is entirely static like this, then you may get random values along the _x_-axis. Let's take 4 ticks. For 100 values, it works fine: 20, 40, 60, 80. However, if you had 93 values, you'd get 18.6, 32.7, 55.8, and 74.4. Is that desired?

**Reply by daniel-eh** [reply]
More or less. To give a little more clarity, an API gives me a dataset with a date that consists of dates and a number associated with each, so the x-axis is the dates and the y-axis is the number. The graph should just show ticks to broadly represent each month, kinda like the picture I attached above, but each point in the dataset still gets plotted. I don't know how many dates will be included since the dataset can vary in size/time covered.

For now what I've done is just hide the graph's x-axis and just draw a horizontal line on a canvas under the graph, and pass in how many months there are and draw equidistant ticks and labels. So it's not actually part of the graph, it's just a fake. This works for now, but I'm hoping to just use the graph to accomplish this rather than having to maintain a separate composable to fake the look.

**Reply by Gowsky** [reply]
Generally, you can achieve this via `HorizontalAxis.ItemPlacer`. We'd need some more information to describe the optimal solution in greater detail.

Could you tell us more about the _x_ values? Are they Unix timestamps? Also, does the data always include full months, or could it be e.g. January 25 to April 17? If the latter is possible, then what, exactly, is the desired behavior in such a case?

**Reply by daniel-eh** [reply]
The timestamps will be a Kotlin `LocalDate` in the format YYYY-MM-DD - for the purposes of a label I would just be grabbing the month from the `LocalDate` and showing it's abbreviation as a tick on the x-axis ("Mar", "Apr", "May", etc). The user doesn't need to see the entire date, the ticks are just to show a generalized change over time. The days will be consecutive as well (2025-01-15, 2025-01-16, 2025-01-17)

**Reply by Gowsky** [reply]
It sounds like the connection between the labels and the data should be relatively loose, meaning that there's supposed to be some number of equally spaced and centered lablels even if this means that:

- They fall in between entries.
- The labels for different months aren't necessarily at the same day for each month.

You can implement `HorizontalAxis.ItemPlacer` and override `getLabelValues` as required. The function receives the _x_ range. Given this and the desired label count, you can easily calculate and return the _x_ values of the labels.

If the label count is dynamic and depends on the _x_ range, then you should update it in a way that ensures synchronization. This means either using extras or simply calculating the label count in `getLabelValues`.
